### PR TITLE
raft: support quorum check when raft is leader

### PR DESCRIPTION
**Pull request by xiang90**
If quorum check fails, the leader will step down to follower.

Fix https://github.com/coreos/etcd/issues/3866

### CODE REVIEW

File: raft/raft.go [hunk 1]
```diff
@@ -381,21 +397,34 @@ func (r *raft) appendEntry(es ...pb.Entry) {
...
 func (r *raft) tickHeartbeat() {
-	r.elapsed++
-	if r.elapsed >= r.heartbeatTimeout {
-		r.elapsed = 0
+	r.heartbeatElapsed++
+	r.electionElapsed++
+
+	if r.electionElapsed >= r.electionTimeout {
+		r.electionElapsed = 0
+		if r.checkQuorum {
```
Comment (gyuho) [inline raft/raft.go]: Curious. Could you explain for me why `r.checkQuorum` instead of `checkQuorumActive`?

Comment (xiang90) [inline raft/raft.go]: just try to make the public configuration variable short. We already have doc for that variable. See line 104. But if you want, we can make them consistent too. Probably we should rename the msg to MsgCheckQuorum

Comment (gyuho) [inline raft/raft.go]: Oh I meant, we have method `checkQuorumActive` and member var `checkQuorum`. Why we don't use the method `checkQuorumActive` here?

And I like your var naming. No complaint :)

Comment (xiang90) [inline raft/raft.go]: If the `checkQuorum` var is true, we will do the checking by sending a msg to state machine or we will not do the checking at all.

The input of the statemachine MUST be a message.

Comment (gyuho) [inline raft/raft.go]: Oh `checkQuorum` var triggers to trigger message sending. Now makes sense. Thanks,

File: raft/raft.go [hunk 2]
```diff
@@ -793,9 +831,36 @@ func (r *raft) loadState(state pb.HardState) {
...
+		}
+
+		if r.prs[id].recentActive {
+			act += 1
+		}
+
+		r.prs[id].recentActive = false
+	}
+
+	if act >= r.q() {
```
Comment (gyuho) [inline raft/raft.go]: maybe just

```
    return act >= r.q()
```

?

Comment (xiang90) [inline raft/raft.go]: oh. yes.

Comment (xiang90) [inline raft/raft.go]: fixed.

Comment (gyuho) [inline raft/raft.go]: Thanks, LGTM (code-wise)

File: raft/raft.go [hunk 3]
```diff
@@ -381,21 +397,34 @@ func (r *raft) appendEntry(es ...pb.Entry) {
...
+	if r.electionElapsed >= r.electionTimeout {
+		r.electionElapsed = 0
+		if r.checkQuorum {
+			r.Step(pb.Message{From: r.id, Type: pb.MsgCheckQuorum})
+		}
+	}
+
+	if r.state != StateLeader {
+		return
+	}
```
Comment (yichengq) [inline raft/raft.go]: what about putting all checkQuorum logic into one if statement?
```
if r.checkQuorum {
  r.electionElapsed++
  if r.electionElapsed >= r.electionTimeout {
    r.Step(pb.Message{From: r.id, Type: pb.MsgCheckQuorum})
  }
  if r.state != StateLeader {
    return
  }
}
```

File: raft/raft.go [hunk 4]
```diff
@@ -157,7 +161,15 @@ type raft struct {
 	// New configuration is ignored if there exists unapplied configuration.
 	pendingConf bool
 
-	elapsed          int // number of ticks since the last msg
+	// number of ticks since it reached last electionTimeout
```
Comment (yichengq) [inline raft/raft.go]: `reached last electionTimeout` is inaccurate IMO. the var could be reset when receiving any message from leader.

File: raft/progress.go [hunk 5]
```diff
@@ -56,6 +56,11 @@ type Progress struct {
 	// is reported to be failed.
 	PendingSnapshot uint64
 
+	// recentActive is true if the progress is recently active. Receiving any messages
+	// from the corresponding follower indicates the progress is active.
+	// recentActive can be reset to false after an election timeout.
+	recentActive bool
```
Comment (yichengq) [inline raft/progress.go]: is name `active` good enough? active always indicates the recent status of the progress.

**Comment by yichengq** [comment]
The approach is good to me.

File: raft/raftpb/raft.proto [hunk 6]
```diff
@@ -45,6 +45,7 @@ enum MessageType {
 	MsgHeartbeatResp   = 9;
 	MsgUnreachable     = 10;
 	MsgSnapStatus      = 11;
+	MsgCheckQuorum     = 12;
```
Comment (bdarnell) [inline raft/raftpb/raft.proto]: This should be added to `IsLocalMsg` in `util.go`.

**Comment by bdarnell** [comment]
LGTM

File: raft/raftpb/raft.proto (hunk 6 above)
Comment (xiang90) [inline raft/raftpb/raft.proto]: fixed.

**Pull request merged**